### DISCUSSION TITLE
Bond orders and optimizer issues

### DISCUSSION THREAD

**Discussion by benedictsaunders** [lukasturcani/stk#296] [Q&A] [answered]
I'm building COFs which have double bonds between the building blocks, but due to the nature of the repeat units, I have had to define by own functional groups. However, using GenericReactionFactory to define the bond orders, it defaults to 1 such that I have trivalent carbons. 

Additionally, both MCHammer and Collapser result in slight distortions to the hexagonal geometry; the bonds remain as long as ever!

Any help or suggestions would be greatly appreciated!

```
    newNHFG = stk.SmartsFunctionalGroupFactory(
        smarts="[C,c]=[N]([H])",
        bonders=(0,),
        deleters=(1, 2),
    )

    bbAldehyde = stk.BuildingBlock(
        smiles=bb1,
        functional_groups=[
            stk.AldehydeFactory(),
        ],
    )

    bbNH = stk.BuildingBlock(
        smiles=bb2,
        functional_groups=[
            newNHFG,
        ],
    )

    cof_stk = stk.ConstructedMolecule(
        topology_graph=stk.cof.Honeycomb(
            building_blocks=(bbNH, bbAldehyde),
            lattice_size=(3, 3, 1),
            reaction_factory=stk.GenericReactionFactory(
                bond_orders={
                    frozenset({newNHFG, stk.Aldehyde}): 2,
                },
            ),
            optimizer=stk.Collapser(
                distance_threshold=1.35,
            ),
        )
    )
    cof_stk.write(f"{str(ID)}_stk.mol")
```

**Comment by lukasturcani** [comment, marked as answer]
Hi @benedictsaunders, welcome to stk! 

### bonder orders

I think the issue here is that in your mapping 

```python
    reaction_factory=stk.GenericReactionFactory(
        bond_orders={
            frozenset({newNHFG, stk.Aldehyde}): 2,
        },
    ),
```

Your set consists of `newNHFG` and `stk.Aldehyde`. The issue is that `newNHFG` is not a `FunctionalGroup` class, it is a `FunctionalGroupFactory` instance, but `bond_oders` expects the set to hold two `FunctionalGroup` classes. The `SmartsFunctionalGroupFactory` creates instances of `GenericFunctionalGroup` so if you do 

```python
    reaction_factory=stk.GenericReactionFactory(
        bond_orders={
            frozenset({stk.GenericFunctionalGroup, stk.Aldehyde}): 2,
        },
    ),
```

it should create double bonds bonds between `GenericFunctionalGroup` instances (ie those created by the `SmartsFunctionalGroupFactory`) and the `Aldehyde` instances. 


### optimizers

For your second question, I think thats becuse I haven't merged this PR by @andrewtarzia (https://github.com/lukasturcani/stk/issues/282) so I will merge it today and you should be able to use features from there from tomorrow.

**Reply by benedictsaunders** [reply]
Hi @lukasturcani, thanks for the prompt response!

I added `GenericFunctionalGroup`, and it works perfectly. I'll be back if I run into a problem with the optimizers 😉 

B

**Reply by lukasturcani** [reply]
welp, I'm busier at work than I expected but worst case I will have the collapser in by next Monday!

**Reply by benedictsaunders** [reply]
Absolutely no stress, thanks for letting me know

B

**Reply by benedictsaunders** [reply]
Sorry to be a pain, do you have an updated estimate on the collapser implementation? I'm just trying to figure out project timelines!

**Reply by lukasturcani** [reply]
Hey, sorry for the delay. Yeah I've managed to get some time to work on the PR yesterday, and I will also have some time today. So hopefully today, but to give myself a more relaxed window let's say next Monday for sure. I'll keep you updated.

**Reply by lukasturcani** [reply]
Hey hey, I think I'm almost done, so expect a new version of stk tomorrow.

**Reply by lukasturcani** [reply]
New release is out, if you're using pip you can run `pip install stk --upgrade` -> the version should be `2021.04.26.0`.

You can make an optimized COF with something like 

```python

bb1 = stk.BuildingBlock('BrCCBr', [stk.BromoFactory()])
bb2 = stk.BuildingBlock('BrCC(CBr)CBr', [stk.BromoFactory()])
cof = stk.ConstructedMolecule(
    topology_graph=stk.cof.PeriodicHoneycomb(
        building_blocks=(bb1, bb2),
        lattice_size=(3, 3, 1),
        optimizer=stk.PeriodicCollapser(),
    ),
)
stk.MolWriter().write(
    molecule=cof,
    path='example.mol',
)


```

**Reply by lukasturcani** [reply]
I just noticed that your example does not use a periodic cof. In this case, I believe the only thing you need to do is 
set the `scale_steps` parameter to `False`.

I.e. taking your example
```python
    cof_stk = stk.ConstructedMolecule(
        topology_graph=stk.cof.Honeycomb(
            building_blocks=(bbNH, bbAldehyde),
            lattice_size=(3, 3, 1),
            reaction_factory=stk.GenericReactionFactory(
                bond_orders={
                   frozenset({stk.GenericFunctionalGroup, stk.Aldehyde}): 2,
                },
            ),
            optimizer=stk.Collapser(
                scale_steps=False,
            ),
        )
    )
```

See the Example section of https://stk.readthedocs.io/en/latest/stk.molecular.topology_graphs.cof.cof.html#stk.molecular.topology_graphs.cof.cof.Cof

**Reply by benedictsaunders** [reply]
Thanks for the update!
